Sources/FirebaseAuth/FirebaseUser+Swift.swift

```diff
@@ -84,10 +85,26 @@ extension User {
     }
   }
 
-  // public mutating func reauthenticate(with credential: AuthCredential) async throws
-  //     -> AuthResult {
-  //   fatalError("\(#function) not yet implemented")
-  // }
+  public mutating func reauthenticate(with credential: Credential) async throws
+      -> AuthResult {
+      typealias Promise = CheckedContinuation<firebase.auth.AuthResult, any Error>
+      return try await withCheckedThrowingContinuation { (continuation: Promise) in
+        let future = self.ReauthenticateAndRetrieveData(credential)
+        withUnsafePointer(to: continuation) { continuation in
+          future.OnCompletion_SwiftWorkaround({ future, pvContinuation in
+            let pContinuation = pvContinuation?.assumingMemoryBound(to: Promise.self)
+            if future.pointee.error() == 0 {
+              pContinuation.pointee.resume(returning: future.pointee.__resultUnsafe().pointee)
+            } else {
+              let code = future.pointee.error()
+              let message = String(cString: future.pointee.__error_messageUnsafe()!)
+              pContinuation.pointee.resume(throwing: FirebaseError(code: code, message: message))
+            }
+          }, UnsafeMutableRawPointer(mutating: continuation))
+        }
+        future.Wait(firebase.FutureBase.kWaitTimeoutInfinite)
```

Comment:
There's no need to call `Wait` on the future here. That just blocks the calling thread needlessly. Registering the callback is sufficient to get notified later when the future resolves. (We may need to fix up other callsites that work similarly as this was the old way we were doing things.) See Query+Swift.swift for the new way.

Also I recommend that we provide the `completion` callback form of this function to better reflect the API surface provided by the Obj C API. That's also what we do in Query+Swift.swift.

Reply:
That all makes sense, however to not grow this scope of this work which is already just a sidetrack to unblock what I'm actually trying to ship, I created the ticket to fix this and added todo to this file.
The rationale is that this follows the style and assumptions of the whole file, and I would rather not go into rabbit hole of the multi threaded multi language environment as a tangent. It looks to me like we should spend some time to adapt the whole file in one go.

https://linear.app/the-browser-company/issue/WPP-1581/refactor-user-class-to-use-new-approach-to-thread-management

Reply:
Got it. I'm also happy to do the cleanup. I found that this pattern was suboptimal when implementing other methods and built the new way that I feel confident works well. I'm happy to make the change.